File: src/main/kotlin/org/skialin/Rect.kt
package org.skialin

data class Rect(
    val left: Float,
    val top: Float,
    val right: Float,
    val bottom: Float,
) {
    val width: Float get() = right - left
    val height: Float get() = bottom - top
    val isEmpty: Boolean get() = !(left < right && top < bottom)

    companion object {
        fun makeWH(
            width: Float,
            height: Float,
        ): Rect = Rect(0f, 0f, width, height)

        fun makeXYWH(
            left: Float,
            top: Float,
            width: Float,
            height: Float,
        ): Rect = Rect(left, top, left + width, top + height)

        fun makeLTRB(
            left: Float,
            top: Float,
            right: Float,
            bottom: Float,
        ): Rect = Rect(left, top, right, bottom)
    }
}

File: rust/skialin-core/src/rect.rs
use crate::sys;

#[derive(Copy, Clone, Debug, Default, PartialEq)]
pub struct Rect {
    pub left: f32,
    pub top: f32,
    pub right: f32,
    pub bottom: f32,
}

impl Rect {
    pub const fn new(left: f32, top: f32, right: f32, bottom: f32) -> Self {
        Rect { left, top, right, bottom }
    }

    pub const fn from_wh(width: f32, height: f32) -> Self {
        Rect::new(0.0, 0.0, width, height)
    }

    pub fn width(&self) -> f32 {
        self.right - self.left
    }

    pub fn height(&self) -> f32 {
        self.bottom - self.top
    }
}

impl From<Rect> for sys::SkRect {
    fn from(r: Rect) -> Self {
        sys::SkRect { fLeft: r.left, fTop: r.top, fRight: r.right, fBottom: r.bottom }
    }
}

impl From<sys::SkRect> for Rect {
    fn from(r: sys::SkRect) -> Self {
        Rect { left: r.fLeft, top: r.fTop, right: r.fRight, bottom: r.fBottom }
    }
}
